### DISCUSSION TITLE
firstOrCreate prioritizes first parameters over second parameters

### DISCUSSION THREAD

**Discussion by Sirfrummel** [laravel/framework#38303] [Ideas] [answered]
- Laravel Version: 8.53.1
- PHP Version: 8.0.9
- Database Driver & Version:  mysql

### Description:

firstOrCreate is prioritizing the FIND parameters _over_ the CREATE parameters, and this seems backwards to me.  For example:
`      $project = Project::firstOrCreate([
          'name' => 'someExistingName',
        ],
        [
          'name' => 'myNewCreateName'
        ]
      );`
I want to say  "find a project by this name, OR create by this other name", and you'll see "someExistingName" is what gets inserted into the database.

I double checked the code and it looks like it's because of how PHP merges 2 arrays, due to this line:
https://github.com/laravel/framework/blame/8.x/src/Illuminate/Database/Eloquent/Builder.php#L484

So if you want to know the actual real world example, of how the _Create_ parameter would be different than the find parameter, take using an object with uuids:

`    $project = Project::firstOrCreate([
          'uuid' => $request->uuid
        ],
        [
          'uuid' => Str::uuid()
        ]
      );
`
So what is happening here, is that it is not over-riding the request UUID with the generated one.  Because basically, if my project doesn't exist with the UUID given by the frontend, I want to generate my own on the backend.  

I feel like it should logically play out like this:
"Use these parameters to FIND the object, if it doesn't exist, use THESE PARAMETERS to CREATE the object"...  it **makes sense** that the FIND parameters should **influence** the CREATE parameters, however, they should not **take precedence** over parameters specifically passed in through the CREATE parameters.

I hope that helps give context and makes sense.  Please tell me if there is something obvious I missed because I am basically a Laravel newbie :)

**Comment by JayBizzle** [comment]
@driesvints are you saying this is expected behaviour rather than a bug? (as the initial issue was converted to a discussion)

Seems more like a bug to me 😕

**Reply by Sirfrummel** [reply]
Yeah I kinda thought this was a bug... but I also thought I was kinda crazy since I'm a newb and it seemed this had been in place for a long time.

**Reply by JayBizzle** [reply]
Yes, seems strange that this hasn't come up before....doesn't mean it isn't wrong though

**Reply by JayBizzle** [reply]
Simple fix is to change this line...

https://github.com/laravel/framework/blob/43ed474c14085125d66cc6b534a654eac43c5160/src/Illuminate/Database/Eloquent/Builder.php#L484-L486

To...

```php
        return tap($this->newModelInstance(array_merge($attributes, $values)), function ($instance) {
            $instance->save();
        });
```

But I guess this would be a breaking change if people are relying on the broken behaviour

**Comment by JayBizzle** [comment]
Same "bug" exists with `firstOrNew()`

```
$company = Company::firstOrNew(
	['name' => 'Does not exist'],
	['name' => 'New Company'],
);

dd($company);

```

Output
```
=> App\Models\Company {
    name: "Does not exist",
}
```

Expected?
```
=> App\Models\Company {
    name: "New Company",
}
```

**Reply by Sirfrummel** [reply]
My understanding is that on the line I referenced in the source code, if the two variables order are switched, it'd have the intended effect.
From `$attributes + $values` to `$values + $attributes`

**Reply by JayBizzle** [reply]
Yeah, that would also do it, although i think the `array_merge()` option gives more clarity

**Reply by Sirfrummel** [reply]
I agree, I didn't realize array merge over writes later keys (had to lookup the docs), I +1 array merge 😃

**Reply by JayBizzle** [reply]
If I were you, i'd put together a PR with the `array_merge()` "fix" along with the details you gave in the first post.

That'll be the best way to get some feedback. If you don't get chance, i'll put together a PR for you, but won't have chance until at least tomorrow evening 👍

**Reply by JayBizzle** [reply]
See https://github.com/laravel/framework/pull/38346

**Reply by JayBizzle** [reply]
Now fixed 👍

**Reply by Sirfrummel** [reply]
@JayBizzle thank you so much for pushing this through.  Can't wait to pull it in (I guess in the next release?)

**Reply by JayBizzle** [reply, marked as answer]
Now released https://github.com/laravel/framework/releases/tag/v8.55.0 👍
